src/odemis/driver/keysight.py

```diff
@@ -459,69 +560,118 @@ def close(self):
+            channel = int(msg[4])
             tracking_mode = msg.split()[1]
-            self._output_states['channel' + channel].tracking = tracking_mode
+            self._output_states[channel].tracking = tracking_mode
         elif re.match(r"sour[1-2]:volt:low ", msg):
-            channel = msg[4]
+            channel = int(msg[4])
             voltage_low = float(msg.split()[1])
-            self._output_states['channel' + channel].voltage_low = voltage_low
+            self._output_states[channel].voltage_low = voltage_low
         elif re.match(r"sour[1-2]:volt:high ", msg):
-            channel = msg[4]
+            channel = int(msg[4])
             voltage_high = float(msg.split()[1])
-            self._output_states['channel' + channel].voltage_high = voltage_high
+            self._output_states[channel].voltage_high = voltage_high
         elif re.match(r"sour[1-2]:volt:offs ", msg):
-            channel = msg[4]
+            channel = int(msg[4])
             voltage_offset = float(msg.split()[1])
-            self._output_states['channel' + channel].voltage_offset = voltage_offset
+            # Sets the average, by keeping the amplitude constant
+            c = self._output_states[channel]
+            vpp = c.voltage_high - c.voltage_low
+            c.voltage_low = voltage_offset - vpp / 2
+            c.voltage_high = voltage_offset + vpp / 2
         elif re.match(r"sour[1-2]:volt:lim:high ", msg):
-            channel = msg[4]
+            channel = int(msg[4])
             voltage_lim_min = float(msg.split()[1])
-            self._output_states['channel' + channel].voltage_limit_low = voltage_lim_min
+            self._output_states[channel].voltage_limit_low = voltage_lim_min
         elif re.match(r"sour[1-2]:volt:lim:low ", msg):
-            channel = msg[4]
+            channel = int(msg[4])
             voltage_lim_max = float(msg.split()[1])
-            self._output_states['channel' + channel].voltage_limit_high = voltage_lim_max
+            self._output_states[channel].voltage_limit_high = voltage_lim_max
         elif re.match(r"sour[1-2]:func:squ:dcyc ", msg):
-            channel = msg[4]
+            channel = int(msg[4])
             duty_cycle = float(msg.split()[1])
-            if duty_cycle < 0.01 or duty_cycle > 99.99:
+            # The device only accepts values between 0.01 and 99.99, but to simulate more limited
+            # range when the frequency is high, we use 20-80%.
+            if duty_cycle < 20 or duty_cycle > 80:
                 logging.warning("The duy cycle value is out of bounds: %s given", duty_cycle)
-            else:
-                self._output_states['channel' + channel].duty_cycle = duty_cycle
-        elif re.match(r"sour[1-2]:volt:offs ", msg):
-            channel = msg[4]
-            voltage_offset = float(msg.split()[1])
-            self._output_states['channel' + channel].voltage_offset = voltage_offset
+                self._error = -222
```

Comment:
I see 3 error values `0, -222, 32`

Is it good to create a custom `errno`?
https://docs.python.org/3/library/errno.html

Reply:
They are not relate to the standard C library error codes. They are the codes from the Keysight device. The manual has the complete list (~300 values). However, as there is a function that can return a "english translation" (`getSystemError()`), I would rather not fill up the driver with these codes. I'll just put a comment next to these numbers in the simulator to explain what they mean.